packages/editor/src/components/block-removal-warnings/index.js

```diff
@@ -0,0 +1,52 @@
+/**
+ * WordPress dependencies
+ */
+
+import { __ } from '@wordpress/i18n';
+import { privateApis as blockEditorPrivateApis } from '@wordpress/block-editor';
+
+/**
+ * Internal dependencies
+ */
+import { unlock } from '../../lock-unlock';
+
+const { BlockRemovalWarningModal } = unlock( blockEditorPrivateApis );
+
+// Prevent accidental removal of certain blocks, asking the user for confirmation first.
+const blockRemovalRules = {
+	'core/query': {
+		message: __(
+			'The Query Loop block displays a list of posts or pages, so removing it will prevent that content from displaying.'
+		),
+	},
+	'core/post-content': {
+		message: __(
+			'Removing the Post Content block will stop your post or page content from displaying on this template.'
+		),
+	},
+	'core/post-template': {
+		message: __(
+			'The Post Template block displays each post or page in a Query Loop, so removing it will stop post content displaying in your query loop.'
+		),
+	},
+	'core/pattern-overrides': {
+		message: __(
+			'Deleting a block with pattern instance overrides can break other blocks on your site that have content linked to it.'
+		),
+		callBack: ( select, clientId, blockNamesForPrompt ) => {
```

Comment:
Passing the `select` across package boundaries in a `callback` feels a bit odd to me as an API. Is there a way we can avoid introducing the callback? I.e. would it be okay for the pattern overrides logic here to live in the component instead 🤔

Reply:
We were doing this in the component, but it was suggested that this would be better a[bstracted as callback in the rules](https://github.com/WordPress/gutenberg/pull/58796#discussion_r1483894219), to avoid complicating the block editor code with [block specific checks/filters](https://github.com/WordPress/gutenberg/pull/58796/files#r1486435447).

My reading of it wouldn't be that we are passing the `select` between package boundaries, it is just a callback that is passed into the block-editor package and the method is called within the `block-editor` package, so the `select` is not available for use in any way within the `editor` package. But happy to hear other's thoughts on this.

Reply:
Thanks Glen, I very well could have been overthinking this part! The main thought I had while looking at is is that I find callback signatures easier to comprehend when the params are fairly simple in nature (like with `clientId`, and `blockNamesForPrompt`), whereas `select` is a more powerful thing to be passed through since it's got all sorts of methods attached. So while the callback is executed in the `block-editor` package, the code in this bit of the `editor` package is doing stuff with a `select` where it isn't immediately obvious where the `select` came from.

It's definitely not a strongly held opinion, though, so happy to go with this if it's what folks prefer!

Reply:
we have time to think about this, so will take another look at how we might be able to refactor it without the `select`  - thanks for the feedback.

Reply:
@andrewserong I think we can simplify it down to [just passing in the block attributes](https://github.com/WordPress/gutenberg/pull/58952/commits/d0f24d2d543085e255e71df952676795a35a1ccf). 

We will just need to have a think about whether a third party might need something other than `blockAttributes` and `blockNamesForPrompt` to work out if they need to add a warning.

Reply:
Nice on. `blockAttributes` and `blockNamesForPrompt` sound like good params to start, to me!